File: models/birthday.py
from models import Database

class Birthday(Database):
    def __init__(self):
        super().__init__()
        self.create()

    def create(self):
        self.execute('''CREATE TABLE IF NOT EXISTS birthdays
                    (USERID INT NOT NULL,
                    MONTH INT NOT NULL,
                    DAY INT NOT NULL,
                    ACTIVE INT NOT NULL)''')
    
    def get(self, user_id):
        data = self.execute("SELECT * FROM birthdays WHERE USERID={}".format(user_id))
        if len(data) == 0:
            return [0, 0]
        else:
            return [data[0][1], data[0][2]]

    def get_multi(self):
        data = self.execute("SELECT USERID FROM birthdays")
        rpkg = []
        for item in data:
            rpkg.append(item[0])
        return rpkg
    
    def set(self, user_id, month, day):
        data = self.execute("SELECT * FROM birthdays WHERE USERID={}".format(user_id))
        if len(data) == 0:
            sql = "INSERT INTO birthdays (USERID, MONTH, DAY, ACTIVE) VALUES ({},{},{},{})".format(user_id, month, day, 1)
        else:
            sql = "UPDATE birthdays SET MONTH={}, DAY={}, ACTIVE={} WHERE USERID={}".format(month, day, 1, user_id)
        self.execute(sql)

    def set_active(self, is_active, user_id):
        if is_active:
            isactive_int = 1
        else:
            isactive_int = 0
        self.execute("UPDATE birthdays SET ACTIVE={} WHERE USERID={}".format(isactive_int, user_id))

    def check(self, month, day):
        data = self.execute("SELECT USERID, ACTIVE FROM birthdays WHERE MONTH={} AND DAY={}".format(month, day))
        if len(data) == 0:
            return []
        else:
            birthday_ids = []
            for item in data:
                if item[1] == 1 or item[1] == None:
                    birthday_ids.append(item[0])
            return birthday_ids

    def remove(self, user_id):
        self.execute("DELETE FROM birthdays WHERE USERID={}".format(user_id))

File: db/birthday.js
const logger = require('../utility/logger');
const Database = require('./database');

class BirthdayDB extends Database {
  constructor() {
    super();
    this.create();
  }

  async create() {
    logger.info('Checking/creating birthdays table');
    await this.execute(`CREATE TABLE IF NOT EXISTS birthdays
            (USERID TEXT NOT NULL,
            MONTH INTEGER NOT NULL,
            DAY INTEGER NOT NULL,
            ACTIVE INTEGER NOT NULL)`);
  }

  async get(userId) {
    logger.info(`Getting birthdays for ${userId} from birthdays table`);
    const data = await this.execute(`SELECT * FROM birthdays WHERE USERID=${userId}`);
    return data.length === 0
      ? [0, 0]
      : data[0];
  }

  async getMulti() {
    logger.info('Getting all birthdays');
    const data = await this.execute('SELECT USERID FROM birthdays');
    return data.map(item => item.USERID);
  }

  async set(userID, month, day) {
    logger.info(`Setting ${userID}'s birthday to ${month}/${day}`);
    const data = await this.execute(`SELECT * FROM birthdays WHERE USERID=${userID}`);
    const sql = data.length === 0
      ? `INSERT INTO birthdays (USERID, MONTH, DAY, ACTIVE) VALUES (${userID}, ${month}, ${day}, ${1})`
      : `UPDATE birthdays SET MONTH=${month}, DAY=${day}, ACTIVE=${1} WHERE USERID=${userID}`;
    await this.execute(sql);
  }

  async setActive(isActive, userID) {
    logger.info('Toggling active in birthdays table');
    const isActiveInt = isActive ? 1 : 0;
    await this.execute(`UPDATE birthdays SET ACTIVE=${isActiveInt} WHERE USERID=${userID}`);
  }

  async check(month, day) {
    logger.info(`Checking if any birthdays on ${month}/${day}`);
    const data = await this.execute(`SELECT USERID, ACTIVE FROM birthdays WHERE MONTH=${month} AND DAY=${day}`);
    const birthdayIDs = [];
    if (data) {
      for (const birthday of data) {
        if (birthday.ACTIVE === 1 || birthday.ACTIVE === null) {
          birthdayIDs.push(birthday.USERID);
        }
      }
    }
    return birthdayIDs;
  }

  async remove(userID) {
    logger.info(`Removing ${userID}'s birthday`);
    await this.execute(`DELETE FROM birthdays WHERE USERID=${userID}`);
  }
}

module.exports = BirthdayDB;